File: python/connection-layer/attestation.py
"""Model attestation — signed proof compute ran the claimed model."""

from __future__ import annotations

import hashlib
import time
from typing import Any

from crypto_wallet import Wallet, verify_signature
from proof_hash import sha256_text


def model_digest(model: str) -> str:
    return sha256_text(model.strip())


def build_attestation(
    wallet: Wallet,
    *,
    task_id: str,
    model: str,
    response: str,
    inference_ms: float,
    prompt_hash: str = "",
) -> dict[str, Any]:
    output_hash = sha256_text(response)
    body = {
        "type": "model_attestation",
        "from": wallet.address,
        "public_key": wallet.public_key_hex,
        "task_id": task_id,
        "model": model,
        "model_hash": model_digest(model),
        "output_hash": output_hash,
        "prompt_hash": prompt_hash,
        "inference_ms": round(inference_ms, 1),
        "timestamp": time.time(),
    }
    return wallet.sign_transaction(body)


def verify_attestation(attestation: dict[str, Any]) -> tuple[bool, str]:
    if attestation.get("type") != "model_attestation":
        return False, "Invalid attestation type"
    if not verify_signature(attestation):
        return False, "Invalid attestation signature"
    model = str(attestation.get("model", ""))
    if attestation.get("model_hash") != model_digest(model):
        return False, "Model hash mismatch"
    return True, "valid"

File: rust/crates/noetis-chain/src/attestation.rs
//! Model attestation — signed proof that compute ran the claimed model.
//! Port of Python `attestation.py`.

use crate::pyjson::{round1, sha256_text};
use crate::wallet::{verify_signature, Wallet};
use serde_json::{json, Value};

pub fn model_digest(model: &str) -> String {
    sha256_text(model.trim())
}

pub fn build_attestation(
    wallet: &Wallet,
    task_id: &str,
    model: &str,
    response: &str,
    inference_ms: f64,
    prompt_hash: &str,
) -> Value {
    let body = json!({
        "type": "model_attestation",
        "from": wallet.address,
        "public_key": wallet.public_key_hex,
        "task_id": task_id,
        "model": model,
        "model_hash": model_digest(model),
        "output_hash": sha256_text(response),
        "prompt_hash": prompt_hash,
        "inference_ms": round1(inference_ms),
        "timestamp": crate::now(),
    });
    wallet.sign_transaction(&body)
}

pub fn verify_attestation(attestation: &Value) -> Result<(), String> {
    if attestation.get("type").and_then(Value::as_str) != Some("model_attestation") {
        return Err("Invalid attestation type".into());
    }
    if !verify_signature(attestation) {
        return Err("Invalid attestation signature".into());
    }
    let model = attestation.get("model").and_then(Value::as_str).unwrap_or("");
    if attestation.get("model_hash").and_then(Value::as_str) != Some(model_digest(model).as_str()) {
        return Err("Model hash mismatch".into());
    }
    Ok(())
}
